ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/validatorcache/ActiveValidatorCache.java

```diff
@@ -0,0 +1,126 @@
+    validatorActiveEpochs = new UInt64[initialSize + VALIDATOR_CACHE_SPARE_CAPACITY][];
+    this.spec = spec;
+  }
+
+  void touch(final UInt64 validatorIndex, final UInt64 epoch) {
+    Preconditions.checkNotNull(validatorIndex);
+    Preconditions.checkNotNull(epoch);
+    LOG.trace("Touch validator {} at epoch {}", validatorIndex, epoch);
+
+    if (validatorIndex.intValue() >= validatorActiveEpochs.length) {
+      grow(validatorIndex);
+    }
+    ensureNodeIsInitialized(validatorIndex);
+    validatorActiveEpochs[validatorIndex.intValue()][epoch.mod(3).intValue()] = epoch;
+  }
+
+  boolean isValidatorSeenAtEpoch(final UInt64 validatorIndex, final UInt64 epoch) {
+    Preconditions.checkNotNull(validatorIndex);
+    Preconditions.checkNotNull(epoch);
+
+    if (validatorActiveEpochs.length <= validatorIndex.intValue()) {
+      LOG.debug(
+          "validator index {} exceeds array size {}", validatorIndex, validatorActiveEpochs.length);
+      return false;
+    }
+
+    return validatorActiveEpochs[validatorIndex.intValue()] != null
+        && epoch.equals(validatorActiveEpochs[validatorIndex.intValue()][epoch.mod(3).intValue()]);
+  }
+
+  int getCacheSize() {
+    return validatorActiveEpochs.length;
+  }
+
+  @VisibleForTesting
+  UInt64[] getValidatorEpochs(final UInt64 validatorIndex) {
+    Preconditions.checkArgument(validatorIndex.intValue() < validatorActiveEpochs.length);
+    return validatorActiveEpochs[validatorIndex.intValue()];
+  }
+
+  private void ensureNodeIsInitialized(final UInt64 validatorIndex) {
+    if (validatorActiveEpochs[validatorIndex.intValue()] == null) {
+      validatorActiveEpochs[validatorIndex.intValue()] = new UInt64[3];
+    }
+  }
+
+  private void grow(final UInt64 requestedValidatorIndex) {
+    final int newSize =
+        requestedValidatorIndex.intValue()
+                < validatorActiveEpochs.length + VALIDATOR_CACHE_SPARE_CAPACITY
+            ? validatorActiveEpochs.length + VALIDATOR_CACHE_SPARE_CAPACITY
+            : requestedValidatorIndex.plus(VALIDATOR_CACHE_SPARE_CAPACITY).intValue();
+    LOG.trace(
+        "Growing ActiveValidatorCache from {} to {} elements",
+        validatorActiveEpochs.length,
+        newSize);
+    validatorActiveEpochs = Arrays.copyOf(validatorActiveEpochs, newSize);
+  }
+
+  @Override
+  public void onBlockImported(final SignedBeaconBlock block) {
+    touch(block.getProposerIndex(), spec.computeEpochAtSlot(block.getSlot()));
+  }
+
+  @Override
+  public void onAttestation(final ValidateableAttestation validateableAttestation) {
+    if (validateableAttestation.getIndexedAttestation().isPresent()) {
+      final IndexedAttestation attestation = validateableAttestation.getIndexedAttestation().get();
+      final UInt64 epoch = spec.computeEpochAtSlot(attestation.getData().getSlot());
+      attestation
+          .getAttesting_indices()
+          .forEach((validatorIndex) -> touch(validatorIndex.get(), epoch));
+    }
```

Comment:
I'd either return early if `getIndexedAttestation().isEmpty()` or use `ifPresent`:
```suggestion
        validateableAttestation
        .getIndexedAttestation()
        .ifPresent(
            attestation -> {
              final UInt64 epoch = spec.computeEpochAtSlot(attestation.getData().getSlot());
              attestation
                  .getAttesting_indices()
                  .forEach((validatorIndex) -> touch(validatorIndex.get(), epoch));
            });
```